### DISCUSSION TITLE
Orthographic View and near/far clipping planes

### DISCUSSION THREAD

**Discussion by esiegel** [visgl/deck.gl#6236] [Q&A] [answered]
I seem to be confused as to the relationship between a layer's Z position, the current Zoom level in an Orthographic View, and the near/far clipping planes.

I wrote up a quick code example to help demonstrate the issue.  There I render a single ScatterPlot Layer with a single circle, rendered in a single Orthographic View.
[CODEPEN](https://codepen.io/esiegel/pen/NWgOgGw)

Things I have noticed:
- When I zoom in super far, around 10, the circle will disappear.  Updating the `far` clipping plane value will make it reappear.
- When the Z position value is set to 0 I can zoom in farther than if it is more negative. 
- From looking at logging, I have also noticed that Deckgl internal viewport's `cameraPosition` changes to near zero as we zoom inward.

I choosing my Z position values with the assumption that by default the camera's eye is at `[0,0,1]` looking negatively down z at `[0,0,0]`.  I would think that values from `[0.9, -999.1]` would be within the clipping planes since the default values are 0.1 and 1000 respectively.

The actual issue I'm trying to solve deals with multiple Orthographic Views where I'm hoping to use differing Z positions to control which layers render on top.  Things get complicated with `getPolygonOffset` with multiple views especially since the `layerIndex` is not global, but resets for each View.

Hopefully this rambling makes some since.

Tl;dr
What is the best way to enforce rendering order when one is using multiple layers in multiple views?  Explicit Z values, which seems to cause issues with clipping planes when zooming, or by customizing the `getPolygonOffset` which seems to only be customizable by layerIndex and not view id.

**Comment by esiegel** [comment, marked as answer]
I think I have figured it out.  At least in part.  

My guess is that with orthographic projections, since item's sizes are fixed even with changes in Z position, that we are scaling the size of the objects to make it zoomed.  Which is why we go beyond the far clipping planes when zoom level gets too high.

This is why the red circle disappears when the zoom level is around Math.log2(1000) ~ 9.6.

I guess the question it brings up, is whether I should be using explicit Z level positions to control layer ordering, or if I should be using the `getPolygonOffset` to do it.

**Reply by Pessimistress** [reply]
You are correct, zoom is applied by scaling up the view matrix. It makes sense that as you zoom in, any small z difference gets more and more visible.

Why do you have overlapping views? If you know which layers will be rendered in which view then you should be able to adjust `layerIndex` in `getPolygonOffset` for a particular layer? If it's a scatterplot (or any other 2d visuals) you can always draw it on top by disabling depth test:

```js
parameters: {
  depthTest: false
}
```

Otherwise, to adjust z according to zoom, you can add a `modelMatrix` prop to the layer that scales z by `1/2**zoom`.

**Reply by esiegel** [reply]
I ended up leaving the Z position at 0 for all layers, and used a custom `getPolygonOffset` that shifted things appropriately.
